main/src/ca/uwaterloo/flix/language/phase/PatternExhaustiveness.scala

```diff
@@ -0,0 +1,381 @@
+        case _ => acc
+      }
+      rules.foldRight(List.empty[List[TypedAst.Pattern]])(defaultRow)
+    }
+
+
+    /**
+      * Computes the set of constructors that appear at the root of the
+      * patterns
+      */
+    def rootCtors(rules: List[List[TypedAst.Pattern]]): List[TypedAst.Pattern] = {
+      def rootCtor(pat: List[TypedAst.Pattern], pats: List[TypedAst.Pattern]) = pat.head match {
+        case _: Pattern.Wild => pats
+        case _: Pattern.Var => pats
+        case tg: Pattern.Tag => tg :: pats
+        case p => p :: pats
+      }
+     rules.foldRight(List.empty[TypedAst.Pattern])(rootCtor)
+    }
+
+    /**
+      * True if ctors is a complete signature for exp. A complete signature is when all constructors of a type are
+      * present. E.g. for
+      *
+      * enum Color {
+      *    case Red,
+      *    case Blue,
+      *    case Yellow
+      *  }
+      *
+      * {Red, Blue, Yellow} is a complete signature, but {Red, Blue} is not. Additionally, {Red, Blue, _} is also not
+      * If the constructors are a complete signature, then they are exhaustive for the type, and we just have to
+      * check that their arguments are also exhaustive
+      *
+      * Wildcards are exhaustive, but we need to do some additional checking in that case (@see DefaultMatrix)
+      *
+      * @param root Root of the expression tree
+      * @param exp The expression to match
+      * @param ctors The ctors that we match with
+      * @return
+      */
+    def completeSig(root: TypedAst.Root, exp: TypedAst.Expression, ctors: List[Pattern]): Boolean = {
+      // Enumerate all the constructors that we need to cover
+      val expCtors: List[String] = exp.tpe match {
+        case Type.Bool => List("True", "False")
+        case Type.Unit => List("Unit")
+        case Type.Enum(enum,_) => root.enums.get(enum).get.cases.toList.map(_._1)
+        case Type.FTuple(_) => List("Tuple")
+        /* For numeric types, we consider them as "infinite" types union
+         * Int = ...| -1 | 0 | 1 | 2 | 3 | ...
+         * The only way we get a match is through a wild. Technically, you could, for example, cover a Char by
+         * having a case for [0 255], but we'll ignore that case for now
+         */
+        case p => return false
+      }
+      /* We cover the needed constructors if there is a wild card in the
+       * root constructor set, or if we match every constructor for the
+       * expression
+       */
+      val ctornames = ctors.map {
+        case TypedAst.Pattern.Tag(_, tg, _, _, _) => tg
+        case p => patName(p)
+      }
+      expCtors.toSet.diff(ctornames.toSet).isEmpty
+    }
+
+    /**
+      * Gets the number of arguments for a constructor, takes into account the "fake constructors"
+      *
+      * @param ctor The constructor to get from
+      * @return The number of arguments for the constructor
+      */
+    def countCtorArgs(ctor: Pattern): Int = ctor match {
```

Comment:
I've manually hard coded some stuff to handle "Constructors" that aren't constructors. I don't like the way that I'm basically hard coding strings, but I can't think of a better way... perhaps some sort of enum.